controllers/goharbor/harbor/loadblance.go

```diff
@@ -0,0 +1,196 @@
+	lbRes, err := r.Controller.AddServiceToManage(ctx, r.GetNotaryServerLB(ctx, harbor), server)
+
+	return NotaryServerLB(lbRes), errors.Wrap(err, "cannot add notary server lb service")
+}
+
+func (r *Reconciler) GetNotaryServerLB(ctx context.Context, harbor *goharborv1alpha2.Harbor) *corev1.Service {
+	if harbor.Spec.Expose.Notary == nil || harbor.Spec.Expose.Notary.LoadBalancer == nil {
+		return nil
+	}
+
+	if !harbor.Spec.Expose.Notary.LoadBalancer.Enable {
+		return nil
+	}
+
+	server := r.NormalizeName(ctx, harbor.GetName(), controllers.NotaryServer.String())
+	tls := harbor.Spec.InternalTLS.GetComponentTLSSpec(r.GetInternalTLSCertificateSecretName(ctx, harbor, harbormetav1.NotaryServerTLS))
+
+	name := fmt.Sprintf("%s-lb", server)
+	namespace := harbor.GetNamespace()
+
+	return &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{{
+				Name:       harbormetav1.NotaryServerAPIPortName,
+				Port:       tls.GetInternalPort(),
+				TargetPort: intstr.FromString(harbormetav1.NotaryServerAPIPortName),
+				Protocol:   corev1.ProtocolTCP,
+			}},
+			Selector: map[string]string{
+				r.Label("name"):      name,
+				r.Label("namespace"): namespace,
+			},
+			Type: corev1.ServiceTypeLoadBalancer,
+		},
+	}
+}
+
+type NotarySignerLB graph.Resource
+
+func (r *Reconciler) AddNotarySignerLB(ctx context.Context, harbor *goharborv1alpha2.Harbor, signer NotarySigner) (NotarySignerLB, error) {
+	lbRes, err := r.Controller.AddServiceToManage(ctx, r.GetNotarySignerLB(ctx, harbor), signer)
+
+	return NotarySignerLB(lbRes), errors.Wrap(err, "cannot add notary signer lb service")
+}
+
+func (r *Reconciler) GetNotarySignerLB(ctx context.Context, harbor *goharborv1alpha2.Harbor) *corev1.Service {
+	if harbor.Spec.Expose.Notary == nil || harbor.Spec.Expose.Notary.LoadBalancer == nil {
+		return nil
+	}
+
+	if !harbor.Spec.Expose.Notary.LoadBalancer.Enable {
+		return nil
+	}
+
+	server := r.NormalizeName(ctx, harbor.GetName(), controllers.NotarySigner.String())
+	name := fmt.Sprintf("%s-lb", server)
+	namespace := harbor.GetNamespace()
+
+	return &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{{
+				Name:       harbormetav1.NotarySignerAPIPortName,
+				Port:       goharborv1alpha2.NotarySignerAPIPort,
+				TargetPort: intstr.FromString(harbormetav1.NotarySignerAPIPortName),
+				Protocol:   corev1.ProtocolTCP,
+			}},
+			Selector: map[string]string{
+				r.Label("name"):      name,
+				r.Label("namespace"): namespace,
+			},
+			Type: corev1.ServiceTypeLoadBalancer,
+		},
+	}
+}
+
+type PortalLB graph.Resource
+
+func (r *Reconciler) AddPortalLB(ctx context.Context, harbor *goharborv1alpha2.Harbor, portal Portal) (NotarySignerLB, error) {
```

Comment:
@Ghostbaby 

It seems functions `addLB` and `getLB` can be general functions that can be used among core,total, notary and singer.

Reply:
If it is hard to do, then keeping it as it was is ok for me.